src/Microsoft.OData.Core/JsonLight/ODataJsonLightReader.cs

```diff
@@ -2256,6 +2259,109 @@ private void EndEntry()
             }
         }
 
+        /// <summary>
+        /// For response de-serialization, restore the null values for properties that are omitted
+        /// by the omit-values=nulls preference header.
+        /// </summary>
+        private void RestoreOmittedNullValues()
+        {
+            // Restore omitted null value only when processing response and when
+            // Preference header omit-values=nulls is specified.
+            if (this.jsonLightInputContext.ReadingResponse
+                && this.jsonLightInputContext.MessageReaderSettings.NullValuesOmitted)
+            {
+                IODataJsonLightReaderResourceState resourceState = this.CurrentResourceState;
+                IEdmStructuredType edmStructuredType = resourceState.ResourceType;
+
+                if (resourceState.SelectedProperties == SelectedPropertiesNode.Empty)
+                {
+                    return;
+                }
+                else
+                {
+                    if (resourceState.Resource != null && edmStructuredType != null)
+                    {
+                        ODataResourceBase resource = resourceState.Resource;
+
+                        IEnumerable<IEdmProperty> selectedProperties;
+
+                        if (resourceState.SelectedProperties == SelectedPropertiesNode.EntireSubtree)
+                        {
+                            selectedProperties = GetSelfAndBaseTypeProperties(edmStructuredType);
+                        }
+                        else
+                        {
+                            // Partial subtree. Combine navigation properties and selected properties at the node with distinct.
+                            selectedProperties =
+                                resourceState.SelectedProperties.GetSelectedNavigationProperties(edmStructuredType)
+                                as IEnumerable<IEdmProperty>;
+                            selectedProperties = selectedProperties.Concat(
+                                resourceState.SelectedProperties.GetSelectedProperties(edmStructuredType))
+                                .Distinct();
+                        }
+
+                        foreach (IEdmProperty currentProperty in selectedProperties)
+                        {
+                            Debug.Assert(currentProperty.Type != null, "currentProperty.Type != null");
+                            if (!currentProperty.Type.IsNullable || currentProperty.PropertyKind == EdmPropertyKind.Navigation)
+                            {
+                                // Skip declared properties that are not null-able types, and declared navigation properties
+                                // which are specified using odata.navigationlink annotations whose absence stand for null values.

```

Comment:
As per other comment; this is not true.  An omitted odata.navigationlink does not imply a null value.  the odata.navigationlink always has a (declared or implied) value; in minimal metadata it is omitted if it matches convention, in full metadata it is always written. this has nothing to do with the value of the navigation property.